design/reboot-interface.md

```diff
@@ -0,0 +1,193 @@
+may be assumed to have been stopped.
+
+To provide a declarative interface for performing a reboot, a new
+BareMetalHostRebootRequest CRD will be created. The spec of this custom
+resource will contain only a reference to a BareMetalHost. The status will
+contain timestamps for when the Host was powered off and on again. Once the
+power off time is present in the status, the host has been stopped. Each
+RebootRequest will be owned by the Host it refers to, so that if the Host is
+deleted all associated RebootRequest CRs will be cleaned up.
+
+The actual power management will be performed by the Host controller. This is
+necessary to avoid race conditions by ensuring that the ``Online`` flag and any
+reboot requests are managed in the same place. The Host controller
+will watch RebootRequest objects and trigger a reconciliation of the linked
+Host when one appears or changes.
+
+For simplicity, we will try to avoid creating a separate controller for the
+RebootRequest at all, but instead perform all operations on it in the course of
+reconciling the BareMetalHost object that it references. However, it may yet
+prove necessary to also create a trivial controller for the purposes of basic
+housekeeping (e.g. setting the owner of the resource).
+
+The creation timestamp of the request resource will be interpreted as the time
+of the reboot request. Requests that predate the last booted time will be
+immediately marked as complete without taking any further action. The Host
+controller will timestamp all pending reboot requests as having completed the
+power off stage whenever any of the following occur:
+
+* The Host has been deprovisioned (in this case the ``lastPoweredOn`` field
+  will be empty)
+* The ``lastPoweredOn`` time is after the time of the reboot request
+* The ``poweredOn`` status is ``false``
+
+Whenever there are pending reboot requests, the Host controller will respond by
+powering off the Host (if it is not already) and timestamping all pending
+requests as having completed the power off stage. Power will then be restored
+only if/when the ``Online`` spec of the Host is true. Once Hosts are powered up
+again, any satisfied reboot requests will be timestamped as having completed
+the power on stage, with this timestamp matching the ``lastPoweredOn`` time of
+the Host.
+
+## Drawbacks
+
+This approach entails a whole new custom resource definition to maintain, and
+the instances of this resource must be managed by the Host controller, which
+adds a degree of complexity.
+
+If large numbers of requests are created but never deleted, searching for
+pending requests may start to consume significant resources.
+
+## Future Enhancements
+
+### Labelling
+
+Adding labels to partially completed and completed requests would allow us to
+exclude them when querying for pending requests. That would eliminate the
+performance concerns associated with allowing old requests to accumulate. This
```

Comment:
Keep in mind that the records will be held in the in-memory cache of the operator. Any performance concerns will be around how quickly the operator can scan those results and retrieve a desired resource. Go is fast, and I think you'd need A LOT of records before scanning is going to have a performance impact. That said, labels are still a fine idea, particularly if any other tool (maybe kubectl) is going to search in the same way.

RAM usage may be more impactful than any performance impact.

Reply:
Yeah, it's likely not a problem in practice. But every Host is going to have to do this scan at least once every 60s, so you can imagine that there'll be a certain scale/longevity beyond which we'll want to do something like this or the automatic deletion.